UefiCpuPkg/Library/MpInitLib/MpLib.c

```diff
@@ -2249,13 +2250,22 @@ MpInitLibInitialize (
   // EfiBootServicesCode to avoid page fault if NX memory protection is enabled.
   //
   CpuMpData->WakeupBufferHigh = AllocateCodeBuffer (ApResetVectorSizeAbove1Mb);
+
+  Cr0.UintN = AsmReadCr0 ();
+  if (Cr0.Bits.PG != 0) {
+    RemoveNxProtection ((EFI_PHYSICAL_ADDRESS)(UINTN)CpuMpData->WakeupBufferHigh, ALIGN_VALUE (ApResetVectorSizeAbove1Mb, EFI_PAGE_SIZE));
+  }
+
   CopyMem (
     (VOID *)CpuMpData->WakeupBufferHigh,
     CpuMpData->AddressMap.RendezvousFunnelAddress +
     CpuMpData->AddressMap.ModeTransitionOffset,
     ApResetVectorSizeAbove1Mb
     );
   DEBUG ((DEBUG_INFO, "AP Vector: non-16-bit = %p/%x\n", CpuMpData->WakeupBufferHigh, ApResetVectorSizeAbove1Mb));
+  if (Cr0.Bits.PG != 0) {
+    ApplyReadOnlyMemoryProtection ((EFI_PHYSICAL_ADDRESS)(UINTN)CpuMpData->WakeupBufferHigh, ALIGN_VALUE (ApResetVectorSizeAbove1Mb, EFI_PAGE_SIZE));

```

Comment:
Can you please name the two functions in a consistent way?
* RemoveNxProtection
* ApplyReadOnlyMemoryProtection

Either they can be:
* RemoveNxProtection
* ApplyRoProtection
or:
* RemoveNonExecutableMemoryProtection
* ApplyReadOnlyMemoryProtection

Reply:
Fixed in latest.

Went with your first suggestion, RemoveNxProtection/ApplyRoProtection